src/server/engine_shard_set.cc

```diff
@@ -70,6 +79,66 @@ ShardMemUsage ReadShardMemUsage(float wasted_ratio) {
   return usage;
 }
 
+// TODO: document this class.
+// Thread safe.
+class RoundRobinSharder {
+ public:
+  void InitThreadLocal() {
+    round_robin_prefix_ = absl::GetFlag(FLAGS_round_robin_prefix);
+
+    if (IsEnabled()) {
+      // 500k entries will consume 1mb per thread, and will allow 100 keys with < 1% collision
+      // probability. Since this has a considerable footprint, we only allocate when enabled.
+      constexpr size_t kRoundRobinSize = 500'000;
+      round_robin_shards_.resize(kRoundRobinSize);
+      std::fill(round_robin_shards_.begin(), round_robin_shards_.end(), kInvalidSid);
+    }
+  }
+
+  bool IsEnabled() const {
+    return !round_robin_prefix_.empty();
+  }
+
+  optional<ShardId> TryGetShardId(string_view key, XXH64_hash_t key_hash) {
+    if (!IsEnabled()) {
+      return nullopt;
+    }
+
+    DCHECK(!round_robin_shards_.empty());
+
+    if (!absl::StartsWith(key, round_robin_prefix_)) {
+      return nullopt;
+    }
+
+    size_t index = key_hash % round_robin_shards_.size();
+    ShardId sid = round_robin_shards_[index];
+
+    if (sid == kInvalidSid) {
+      shard_set->pool()->Await([this, index](ProactorBase*) {
+        round_robin_shards_[index] = next_shard_;
+        next_shard_ = (next_shard_ + 1) % shard_set->size();
+      });
+
+      // We must read again from the vector because 2 calls in parallel with the same index might
+      // overwrite each other.
+      sid = round_robin_shards_[index];
```

Comment:
Yes, this is a good catch. I need assignment atomicity here. I'll add that.

Reply:
I think vlad means something else. Basically, `Await` can be called by different threads. Let's say they both arrive at the same time on `Await()`. Both write at index 5. Now the second threads await finishes, making `next_shard_ = 6`. Then the first thread executes the `Await()` and overwrites the `round_robin_shards_[index]` with a different id. 

Now both of the functions move on, and they return a differnt `sid` for the key. 

I am not sure what you mean assignment atomicity here, but the operation itself (anything between the if) should be done "atomically".

P.s. the reason that this can happen is `Await` uses a BlockingCounter which can preempt the active fiber. So when this happens, the other' threads await might run first, followed by the first thread await causing the two threads see a different sid

Reply:
Yes this is a very detailed version of what I meant 😆

Reply:
Thanks Kostas, that's what I meant :)

Vlad, I don't think I fully understand your "liveness" concern. You mean that we will accidentally call `Shard()`, which now may yield, in a non-preeemtable context?

Reply:
Exactly the opposite: Because we now have so many short circuit paths, the Shard() call can spend a long time waiting for some other thread to finally preempt and let it execute there (Remember the bug with replicaof not called under heavy load). But again as its called rarely we can likely neglect it and solve it if it becomes a problem

Reply:
another option would be to maintain another, "global" table that would be source of truth and protect it with the spinlock. 
then each thread updates or reads only from it.

Reply:
That requires reading an atomic 🙄 

A no-lock / no-hop approach would be keeping an update queues per shard. Once we find an undefined shard id, we first consume our local update queue, and if its still undefined, push updates to other shards under a critical section. The local update queue can be just anything we have: lockfree or lockless. In the end the slow path becomes more complicated, but the hot path is dead simple and we don't need any hops ever. 

But again, my problem is unlikely to occur in our current use case

Reply:
I meant to maintain the global table in addition to thread-local copy. Similar design to lua scripts manager.
The threads access shared memory only if they do not have any info in their local tables. But yes, we agree at high-level: hot path is dead simple.

Reply:
Ah yes, that's actually simpler

Reply:
Nice idea, done.